runtime/major_gc.c

```diff
@@ -344,45 +344,122 @@ static uintnat default_slice_budget() {
-  if ( Caml_state->opportunistic_work > computed_work ) {
-    /* bound the credit that opportunistic_work can have vs computed_work */
-    if (Caml_state->opportunistic_work > 2*computed_work) {
-     Caml_state->opportunistic_work = 2*computed_work;
-    }
+  /* accumulate work */
+  dom_st->major_work_computed += computed_work;
+  dom_st->major_work_todo += computed_work;
 
-    opportunistic_credit = computed_work;
-    Caml_state->opportunistic_work -= computed_work;
-    computed_work = 0;
-  } else {
-    opportunistic_credit = Caml_state->opportunistic_work;
-    Caml_state->opportunistic_work = 0;
-    computed_work -= opportunistic_credit;
+  /* cap accumulated work todo to p = 0.3 */
+  limit = (intnat)(0.3 * (heap_sweep_words + (heap_words * 100 / (100 + caml_percent_free))));
+  if (dom_st->major_work_todo > limit)
+  {
+    dom_st->major_work_todo = limit;
   }
 
   caml_gc_message (0x40, "heap_words = %"
                          ARCH_INTNAT_PRINTF_FORMAT "u\n",
-                         (uintnat)heap_words);
+                   (uintnat)heap_words);
   caml_gc_message (0x40, "allocated_words = %"
                          ARCH_INTNAT_PRINTF_FORMAT "u\n",
-                   Caml_state->allocated_words);
-  caml_gc_message (0x40, "amount of work to do = %"
+                   dom_st->allocated_words);
+  caml_gc_message (0x40, "raw work-to-do = %"
                          ARCH_INTNAT_PRINTF_FORMAT "uu\n",
                    (uintnat) (p * 1000000));
-  caml_gc_message (0x40, "ordered work = %ld words\n", (intnat)-1);
-  caml_gc_message (0x40, "opportunistic work credit = %ld words\n", opportunistic_credit);
-  caml_gc_message (0x40, "computed work = %ld words\n", computed_work);
+  caml_gc_message (0x40, "computed work = %"
+                         ARCH_INTNAT_PRINTF_FORMAT "d words\n",
+                   computed_work);
+
+  caml_gc_log("Updated major work: [%c] "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "u heap_words, "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "u allocated, "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "d computed_work, "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "d work_computed, "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "d work_todo, "
+                         " %"ARCH_INTNAT_PRINTF_FORMAT "u gc_clock",
+                         caml_gc_phase_char(caml_gc_phase),
+                         (uintnat)heap_words, dom_st->allocated_words,
+                         computed_work,
+                         dom_st->major_work_computed,
+                         dom_st->major_work_todo,
+                         (intnat)(dom_st->major_gc_clock*1000000));
+
+  dom_st->stat_major_words += dom_st->allocated_words;
+  dom_st->allocated_words = 0;
+}
+
+static intnat get_major_slice_work(intnat howmuch) {
+  caml_domain_state *dom_st = Caml_state;
+  intnat computed_work;
+
+  /* calculate how much work to do now */
+  if (howmuch == -1) {
```

Comment:
Combine if and else if branches into `howmuch == -1 || howmuch == 0`.

Reply:
done!